File: killer_sudoku/solver/engine/board_state.py
"""BoardState — all mutable solver state, single mutation point.

BoardState is constructed from a PuzzleSpec and holds:
- candidates[r][c]: set of remaining digit candidates per cell (0-based)
- counts[unit_id][digit]: how many cells in that unit still have digit as candidate
- unit_versions[unit_id]: increments on every candidate removal in that unit
- cage_solns[cage_idx]: list of remaining valid frozenset digit assignments
- units: list of Unit objects (rows 0-8, cols 9-17, boxes 18-26, cages 27+)
- regions[r][c]: 0-based cage index for cell (r, c)

Global unit ID layout:
  rows    0..8
  cols    9..17
  boxes   18..26
  cages   27..27+n_cages-1
  virtual 27+n_cages..
"""

from __future__ import annotations

import dataclasses

import numpy as np
import numpy.typing as npt

from killer_sudoku.solver.engine.linear_system import LinearSystem
from killer_sudoku.solver.engine.types import BoardEvent, Cell, Trigger, Unit, UnitKind
from killer_sudoku.solver.equation import sol_sums
from killer_sudoku.solver.puzzle_spec import PuzzleSpec


class NoSolnError(Exception):
    """Raised when a cell's candidate set would become empty."""


def _box_cells(box: int) -> frozenset[Cell]:
    r0, c0 = (box // 3) * 3, (box % 3) * 3
    return frozenset((r0 + dr, c0 + dc) for dr in range(3) for dc in range(3))


@dataclasses.dataclass
class BoardState:
    """All mutable solver state. Constructed from a validated PuzzleSpec.

    Rules read from this object but must never mutate it directly.
    All mutations go through remove_candidate() or remove_cage_solution().
    """

    spec: PuzzleSpec
    units: list[Unit]
    candidates: list[list[set[int]]]  # [9][9]
    counts: list[list[int]]  # [n_units][10]
    unit_versions: list[int]  # [n_units]
    cage_solns: list[list[frozenset[int]]]  # [n_cages][*]
    regions: npt.NDArray[np.intp]  # (9,9) 0-based cage index
    _cell_unit_ids: list[list[list[int]]]  # [9][9] -> list of unit_ids
    linear_system: LinearSystem

    def __init__(self, spec: PuzzleSpec, *, include_virtual_cages: bool = True) -> None:
        self.spec: PuzzleSpec = spec
        # Convert regions to 0-based
        self.regions = spec.regions - 1
        n_cages = int(self.regions.max()) + 1

        # Build cage cell lists (0-based cage index)
        cage_cells_list: list[list[Cell]] = [[] for _ in range(n_cages)]
        for r in range(9):
            for c in range(9):
                cage_cells_list[int(self.regions[r, c])].append((r, c))

        # Build unit list: rows 0-8, cols 9-17, boxes 18-26, real cages 27..
        self.units: list[Unit] = []
        for r in range(9):
            self.units.append(
                Unit(r, UnitKind.ROW, frozenset((r, c) for c in range(9)))
            )
        for c in range(9):
            self.units.append(
                Unit(9 + c, UnitKind.COL, frozenset((r, c) for r in range(9)))
            )
        for b in range(9):
            self.units.append(Unit(18 + b, UnitKind.BOX, _box_cells(b)))
        for idx, cells in enumerate(cage_cells_list):
            self.units.append(Unit(27 + idx, UnitKind.CAGE, frozenset(cells)))

        # Real cage solutions via sol_sums
        self.cage_solns: list[list[frozenset[int]]] = []
        for cells in cage_cells_list:
            total = 0
            for r, c in cells:
                v = int(spec.cage_totals[r, c])
                if v != 0:
                    total = v
                    break
            self.cage_solns.append(sol_sums(len(cells), 0, total))

        # LinearSystem: build now so virtual_cages are available
        self.linear_system = LinearSystem(spec)

        # Add virtual cage units from the linear system (derived sum equations).
        # Virtual cage unit IDs start at 27 + n_cages and are indexed in
        # cage_solns at offset n_cages (so cage_idx = unit_id - 27 works uniformly).
        # Burb virtual cages (distinct=True, precomp_solns=None): sol_sums is correct.
        # Non-burb virtual cages (distinct=False, precomp_solns=list): use the
        # reduce_equns-propagated solutions directly — sol_sums would wrongly
        # assume digit distinctness for cells spanning multiple units.
        # When include_virtual_cages=False (playing mode without LinearElimination
        # active), skip these so the linear system's derived constraints do not
        # leak into candidate computation via CageIntersection.
        for vcells, vtotal, distinct, precomp_solns in (
            self.linear_system.virtual_cages if include_virtual_cages else []
        ):
            vunit_id = len(self.units)
            self.units.append(Unit(vunit_id, UnitKind.CAGE, vcells, distinct))
            if precomp_solns is not None:
                self.cage_solns.append(precomp_solns)
            else:
                self.cage_solns.append(sol_sums(len(vcells), 0, vtotal))

        n_units = len(self.units)

        # Per-cell unit ID lookup — built after all units (incl. virtual cages)
        self._cell_unit_ids = [[[] for _ in range(9)] for _ in range(9)]
        for unit in self.units:
            for r, c in unit.cells:
                self._cell_unit_ids[r][c].append(unit.unit_id)

        # Candidates: start full
        self.candidates = [[set(range(1, 10)) for _ in range(9)] for _ in range(9)]

        # Counts: each digit appears in all cells of each unit initially
        self.counts = [[0] * 10 for _ in range(n_units)]
        for unit in self.units:
            for d in range(1, 10):
                self.counts[unit.unit_id][d] = len(unit.cells)

        self.unit_versions = [0] * n_units

    # --- Unit ID accessors ---

    def row_unit_id(self, r: int) -> int:
        """Unit ID for row r (0-based)."""
        return r

    def col_unit_id(self, c: int) -> int:
        """Unit ID for column c (0-based)."""
        return 9 + c

    def box_unit_id(self, r: int, c: int) -> int:
        """Unit ID for the 3×3 box containing cell (r, c)."""
        return 18 + (r // 3) * 3 + (c // 3)

    def cage_unit_id(self, r: int, c: int) -> int:
        """Unit ID for the cage containing cell (r, c)."""
        return 27 + int(self.regions[r, c])

    def cell_unit_ids(self, r: int, c: int) -> list[int]:
        """All four unit IDs for cell (r, c): row, col, box, cage."""
        return self._cell_unit_ids[r][c]

    # --- Mutation ---

    def remove_candidate(self, r: int, c: int, d: int) -> list[BoardEvent]:
        """Remove digit d from candidates[r][c]; update counts, versions, emit events.

        This is the single mutation point for candidate sets. It:
        1. Removes d from candidates[r][c]
        2. Decrements counts[unit_id][d] for all units containing (r, c)
        3. Emits COUNT_DECREASED, COUNT_HIT_TWO, COUNT_HIT_ONE as counts change
        4. Emits CELL_DETERMINED if candidates[r][c] becomes a singleton
        5. Prunes cage solutions (real and virtual) that are now impossible
        6. Raises NoSolnError if candidates[r][c] would become empty

        Returns a list of BoardEvent objects for the engine to route.
        """
        cands = self.candidates[r][c]
        if d not in cands:
            return []
        if len(cands) == 1:
            raise NoSolnError(f"Cannot remove last candidate {d} from ({r},{c})")

        cands.discard(d)
        events: list[BoardEvent] = []

        for uid in self.cell_unit_ids(r, c):
            prev = self.counts[uid][d]
            new = prev - 1
            self.counts[uid][d] = new
            self.unit_versions[uid] += 1
            events.append(BoardEvent(Trigger.COUNT_DECREASED, uid, d))
            if new == 2:
                events.append(BoardEvent(Trigger.COUNT_HIT_TWO, uid, d))
            elif new == 1:
                events.append(BoardEvent(Trigger.COUNT_HIT_ONE, uid, d))

        if len(cands) == 1:
            sole = next(iter(cands))
            events.append(BoardEvent(Trigger.CELL_DETERMINED, (r, c), sole))

        # Prune solutions for all cage units containing this cell (real + virtual)
        for uid in self.cell_unit_ids(r, c):
            if self.units[uid].kind == UnitKind.CAGE:
                cage_idx = uid - 27
                events.extend(self._prune_cage_solutions(cage_idx, r, c, d))

        return events

    def remove_cage_solution(
        self, cage_idx: int, solution: frozenset[int]
    ) -> BoardEvent:
        """Remove solution by value from cage_solns[cage_idx]; return SOLUTION_PRUNED.

        Called exclusively by _prune_cage_solutions. Rules must not call this
        directly — all mutations are mediated through remove_candidate.
        """
        self.cage_solns[cage_idx].remove(solution)
        cage_unit_id = 27 + cage_idx
        return BoardEvent(Trigger.SOLUTION_PRUNED, cage_unit_id, None)

    def _prune_cage_solutions(
        self, cage_idx: int, r: int, c: int, d: int
    ) -> list[BoardEvent]:
        """Remove cage solutions impossible given current candidates.

        When digit d has been removed from cell (r, c) in this cage, check
        whether any cage solutions now require d in a cell that no longer
        has d as a candidate. If every cell in the cage lacks d, prune all
        solutions containing d (coarse filter; fine-grained per-cell filtering
        is handled by SolutionMapFilter R4).
        """
        cage_unit = self.units[27 + cage_idx]
        if any(d in self.candidates[cr][cc] for cr, cc in cage_unit.cells):
            return []  # d is still possible somewhere in the cage
        # d is impossible in this cage — remove all solutions containing it
        to_remove = [s for s in self.cage_solns[cage_idx] if d in s]
        return [self.remove_cage_solution(cage_idx, s) for s in to_remove]

    def add_virtual_cage(
        self,
        cells: frozenset[Cell],
        total: int,
        eliminated_solns: list[frozenset[int]],
        *,
        distinct: bool = True,
    ) -> None:
        """Add a user-acknowledged virtual cage as a new cage unit.

        Computes initial solutions via sol_sums, removes eliminated_solns,
        and updates units, cage_solns, counts, unit_versions, and
        _cell_unit_ids to include the new unit.

        Args:
            cells:           0-based (row, col) pairs for the cage cells.
            total:           The cage sum constraint.
            eliminated_solns: Solution sets to exclude from cage_solns.
            distinct:        Whether the cage requires distinct digits (default True).
        """
        vunit_id = len(self.units)
        self.units.append(Unit(vunit_id, UnitKind.CAGE, cells, distinct))

        elim_sets = {frozenset(s) for s in eliminated_solns}
        solns = sol_sums(len(cells), 0, total)
        self.cage_solns.append([s for s in solns if s not in elim_sets])

        counts_row = [0] * 10
        for d in range(1, 10):
            counts_row[d] = sum(1 for r, c in cells if d in self.candidates[r][c])
        self.counts.append(counts_row)
        self.unit_versions.append(0)

        for r, c in cells:
            self._cell_unit_ids[r][c].append(vunit_id)


def validate_solution(bs: BoardState) -> list[str]:
    """Validate a fully-solved board against all killer sudoku rules.

    Returns a list of violation strings (empty if the solution is valid).
    Checks:
    - Every cell is determined (singleton candidate set)
    - Each row, column, and 3×3 box contains each digit 1-9 exactly once
    - Each real cage sums to its declared total and uses distinct digits
    """
    violations: list[str] = []

    # All cells must be determined
    for r in range(9):
        for c in range(9):
            if len(bs.candidates[r][c]) != 1:
                violations.append(
                    f"cell ({r},{c}) not determined: {bs.candidates[r][c]}"
                )

    if violations:
        return violations  # Further checks need singleton candidates

    val = [[next(iter(bs.candidates[r][c])) for c in range(9)] for r in range(9)]

    # Rows
    for r in range(9):
        if sorted(val[r]) != list(range(1, 10)):
            violations.append(f"row {r} invalid: {val[r]}")

    # Columns
    for c in range(9):
        col = [val[r][c] for r in range(9)]
        if sorted(col) != list(range(1, 10)):
            violations.append(f"col {c} invalid: {col}")

    # Boxes
    for b in range(9):
        r0, c0 = (b // 3) * 3, (b % 3) * 3
        box = [val[r0 + dr][c0 + dc] for dr in range(3) for dc in range(3)]
        if sorted(box) != list(range(1, 10)):
            violations.append(f"box {b} invalid: {box}")

    # Real cages: sum and distinct digits
    n_real = int(bs.regions.max()) + 1
    for cage_idx in range(n_real):
        unit = bs.units[27 + cage_idx]
        cage_vals = [val[r][c] for r, c in unit.cells]
        if len(cage_vals) != len(set(cage_vals)):
            violations.append(f"cage {cage_idx} has repeated digits: {cage_vals}")
        # Get declared total from cage_totals
        expected_total = 0
        for r, c in unit.cells:
            t = int(bs.spec.cage_totals[r, c])
            if t != 0:
                expected_total = t
                break
        if expected_total and sum(cage_vals) != expected_total:
            violations.append(
                f"cage {cage_idx} sums to {sum(cage_vals)}, expected {expected_total}"
            )

    return violations


def apply_initial_eliminations(bs: BoardState) -> list[BoardEvent]:
    """Apply LinearSystem initial eliminations and return all fired events.

    Called once after BoardState construction to propagate cells that the
    linear system determined at setup time (e.g. single-cell cages).
    """
    all_events: list[BoardEvent] = []
    for elim in bs.linear_system.initial_eliminations:
        r, c = elim.cell
        if elim.digit in bs.candidates[r][c] and len(bs.candidates[r][c]) > 1:
            all_events.extend(bs.remove_candidate(r, c, elim.digit))
    return all_events

File: web/src/engine/boardState.ts
/**
 * BoardState — plain row/col/box sudoku skeleton shared by classic and killer
 * boards. KillerBoardState (below) extends it with cage modelling.
 *
 * The classic/killer split keeps classic boards free of cage machinery (see
 * `docs/superpowers/specs/2026-06-07-cage-free-board-state-for-classic.md`).
 *
 * Rules read from this object but must never mutate it directly.
 * All mutations go through removeCandidate() (or, on KillerBoardState,
 * removeCageSolution()).
 */

import { solSums, solDiffs } from '../solver/equation.js';
import type { DiffSolution } from '../solver/equation.js';
import { NoSolnError } from '../solver/errors.js';
import type { PuzzleSpec } from '../solver/puzzleSpec.js';
import type { CageConstraints } from './backtracker.js';
import { LinearSystem } from './linearSystem.js';
import {
  BoardEvent,
  Cell,
  Elimination,
  Trigger,
  Unit,
  UnitKind,
} from './types.js';

// ---------------------------------------------------------------------------
// Unit layout constants: units[] is partitioned as rows / cols / boxes / cages
// ---------------------------------------------------------------------------

export const ROW_UNIT_OFFSET = 0;   // indices 0–8
export const COL_UNIT_OFFSET = 9;   // indices 9–17
export const BOX_UNIT_OFFSET = 18;  // indices 18–26
export const CAGE_UNIT_OFFSET = 27; // indices 27+

// ---------------------------------------------------------------------------
// Precomputed box cell lists: boxCells[b] = all cells in box b (0-based)
// ---------------------------------------------------------------------------

function buildBoxCells(): readonly (readonly Cell[])[] {
  return Array.from({length: 9}, (_, b) => {
    const cells: Cell[] = [];
    for (let dr = 0; dr < 3; dr++)
      for (let dc = 0; dc < 3; dc++)
        cells.push([(b / 3 | 0) * 3 + dr, (b % 3) * 3 + dc] as Cell);
    return cells;
  });
}

const BOX_CELLS: readonly (readonly Cell[])[] = buildBoxCells();

// ---------------------------------------------------------------------------
// KillerBoardState
// ---------------------------------------------------------------------------

export class BoardState {
  readonly units: Unit[];
  /** candidates[r][c] = set of remaining digits for cell (r, c). Use cands(r,c) for safe read access. */
  candidates: Set<number>[][];
  /** counts[unitId][digit] = number of cells in that unit still having digit. Use count(uid,d) for safe read access. */
  counts: number[][];
  unitVersions: number[];

  private _cellUnitIds: number[][][]; // [9][9] → list of unit_ids

  constructor() {
    this.units = [];
    this._cellUnitIds = Array.from({length: 9}, () => Array.from({length: 9}, () => []));
    this.candidates = Array.from({length: 9}, () =>
      Array.from({length: 9}, () => new Set(Array.from({length: 9}, (_, i) => i + 1))));
    this.counts = [];
    this.unitVersions = [];

    for (let r = 0; r < 9; r++)
      this._addUnit({ unitId: ROW_UNIT_OFFSET + r, kind: UnitKind.ROW,
        cells: Array.from({length: 9}, (_, c) => [r, c] as Cell), distinctDigits: true });
    for (let c = 0; c < 9; c++)
      this._addUnit({ unitId: COL_UNIT_OFFSET + c, kind: UnitKind.COL,
        cells: Array.from({length: 9}, (_, r) => [r, c] as Cell), distinctDigits: true });
    for (let b = 0; b < 9; b++)
      this._addUnit({ unitId: BOX_UNIT_OFFSET + b, kind: UnitKind.BOX, cells: BOX_CELLS[b]!, distinctDigits: true });
  }

  // ── Safe read accessors (invariant: 9×9 board and nUnits always initialised) ─

  /** Candidates for cell (r, c). Indices are always 0–8 by solver invariant. */
  cands(r: number, c: number): Set<number> { return this.candidates[r]![c]!; }

  /** Count of cells in unit uid that still have digit d (d ∈ [1..9]; index 0 is unused). */
  count(uid: number, d: number): number { return this.counts[uid]![d]!; }

  // ── Unit ID accessors ────────────────────────────────────────────────────

  rowUnitId(r: number): number { return ROW_UNIT_OFFSET + r; }
  colUnitId(c: number): number { return COL_UNIT_OFFSET + c; }
  boxUnitId(r: number, c: number): number { return BOX_UNIT_OFFSET + (r / 3 | 0) * 3 + (c / 3 | 0); }
  cellUnitIds(r: number, c: number): number[] { return this._cellUnitIds[r]![c]!; }

  /**
   * Returns eliminations to apply to all peers of (r, c) when it is determined
   * to hold digit d — i.e. d removed from every other cell sharing a row, col,
   * box, or distinct-digit cage with (r, c).
   *
   * Used by NakedSingle and by buildEngine for unconditional placement propagation.
   */
  peerEliminations(r: number, c: number, d: number): Elimination[] {
    const seen = new Set<string>();
    const elims: Elimination[] = [];
    for (const uid of this.cellUnitIds(r, c)) {
      const unit = this.units[uid]!;
      if (unit.kind === UnitKind.CAGE && !unit.distinctDigits) continue;
      for (const [pr, pc] of unit.cells as Cell[]) {
        if (pr === r && pc === c) continue;
        const key = `${pr},${pc}`;
        if (seen.has(key)) continue;
        seen.add(key);
        if (this.cands(pr, pc).has(d)) elims.push({ cell: [pr, pc] as Cell, digit: d });
      }
    }
    return elims;
  }

  // ── Mutation ─────────────────────────────────────────────────────────────

  /**
   * Remove digit d from candidates[r][c]; update counts, versions, emit events.
   *
   * This is the single mutation point for candidate sets. Steps:
   *  1. Remove d from candidates[r][c]
   *  2. Decrement counts[unitId][d] for all units containing (r, c)
   *  3. Emit COUNT_DECREASED / COUNT_HIT_TWO / COUNT_HIT_ONE as counts change
   *  4. Emit CELL_DETERMINED if candidates[r][c] becomes a singleton
   *  5. Raise NoSolnError if candidates[r][c] would become empty
   *
   * KillerBoardState overrides this to additionally prune cage solutions —
   * the same template-method shape as cageConstraints() below.
   */
  removeCandidate(r: number, c: number, d: number): BoardEvent[] {
    const cands = this.cands(r, c);
    if (!cands.has(d)) return [];
    if (cands.size === 1) throw new NoSolnError(`Cannot remove last candidate ${d} from (${r},${c})`);

    cands.delete(d);
    const events: BoardEvent[] = [];

    for (const uid of this.cellUnitIds(r, c)) {
      const prev = this.count(uid, d);
      const next = prev - 1;
      this.counts[uid]![d] = next;
      this.unitVersions[uid]!++;
      events.push({ trigger: Trigger.COUNT_DECREASED, payload: uid, hintDigit: d });
      if (next === 2) events.push({ trigger: Trigger.COUNT_HIT_TWO, payload: uid, hintDigit: d });
      else if (next === 1) events.push({ trigger: Trigger.COUNT_HIT_ONE, payload: uid, hintDigit: d });
    }

    if (cands.size === 1) {
      const sole = nextInSet(cands);
      events.push({ trigger: Trigger.CELL_DETERMINED, payload: [r, c] as Cell, hintDigit: sole });
    }

    return events;
  }

  /**
   * Reduce candidates[r][c] to exactly the given set for every cell, removing
   * everything else via removeCandidate (so KillerBoardState's cage-solution
   * pruning runs correctly). Used to replay a RuleBugFixture's stalledCandidates
   * onto a freshly constructed board.
   */
  restoreCandidates(candidates: readonly (readonly (readonly number[])[])[]): void {
    for (let r = 0; r < 9; r++) {
      for (let c = 0; c < 9; c++) {
        const keep = new Set(candidates[r]![c]!);
        for (const d of [...this.cands(r, c)]) {
          if (!keep.has(d)) this.removeCandidate(r, c, d);
        }
      }
    }
  }

  /**
   * Cage-sum data for the MRV backtracker's validity check, or null when this
   * board has no cages. Plain BoardState always returns null — mrvBacktrack's
   * search() then degrades to pure row/col/box backtracking (cageTotal empty
   * ⟹ cageValid's `if (total === undefined) return true` short-circuit).
   */
  /**
   * Cells outside the standard row/col/box that share an extra distinct-digit
   * constraint with (r, c) — e.g. Big Apple window peers. Empty for a plain
   * board; overridden by BigAppleBoardState.
   */
  extraPeers(_r: number, _c: number): readonly Cell[] { return []; }

  cageConstraints(): CageConstraints | null { return null; }

  /**
   * Register a new unit: append it to `units` and extend `counts` /
   * `unitVersions` / the per-cell unit-ID lookup to cover it.
   *
   * This is the single place that keeps those three parallel arrays in sync
   * with `units` — shared by the row/col/box construction above,
   * KillerBoardState's cage construction, and addVirtualCage.
   */
  protected _addUnit(unit: Unit): void {
    this.units.push(unit);
    const countsRow = new Array<number>(10).fill(0);
    for (let d = 1; d <= 9; d++)
      countsRow[d] = unit.cells.filter(([r, c]) => this.cands(r, c).has(d)).length;
    this.counts.push(countsRow);
    this.unitVersions.push(0);
    for (const [r, c] of unit.cells) this._cellUnitIds[r]![c]!.push(unit.unitId);
  }
}

// ---------------------------------------------------------------------------
// KillerBoardState — adds cage modelling: cage units, cage-solution tracking,
// the linear system, and virtual cages derived from it
// ---------------------------------------------------------------------------

export class KillerBoardState extends BoardState {
  readonly spec: PuzzleSpec;
  /** regions[r][c] = 0-based cage index */
  readonly regions: number[][];
  /** cage_solns[cage_idx] = remaining feasible digit sets for that cage */
  cageSolns: number[][][];
  readonly linearSystem: LinearSystem;

  constructor(spec: PuzzleSpec, { includeVirtualCages = true } = {}) {
    super();
    this.spec = spec;

    // Convert regions to 0-based (spec uses 1-based cage IDs)
    // Indices are loop-bounded 0–8, so [r]![c]! are always valid.
    this.regions = Array.from({length: 9}, (_, r) =>
      Array.from({length: 9}, (__, c) => spec.regions[r]![c]! - 1));
    const nCages = Math.max(...this.regions.flat()) + 1;

    // Build cage cell lists (0-based index)
    const cageCellsList: Cell[][] = Array.from({length: nCages}, () => []);
    for (let r = 0; r < 9; r++)
      for (let c = 0; c < 9; c++)
        cageCellsList[this.regions[r]![c]!]!.push([r, c] as Cell);

    // Real cage units (27+)
    for (let idx = 0; idx < nCages; idx++)
      this._addUnit({ unitId: CAGE_UNIT_OFFSET + idx, kind: UnitKind.CAGE, cells: cageCellsList[idx]!, distinctDigits: true });

    // Real cage solutions via sol_sums
    this.cageSolns = cageCellsList.map(cells => {
      let total = 0;
      for (const [r, c] of cells) {
        const v = spec.cageTotals[r]![c]!;
        if (v !== 0) { total = v; break; }
      }
      return solSums(cells.length, 0, total);
    });

    // Build LinearSystem (this is the expensive step)
    this.linearSystem = new LinearSystem(spec, { deriveVirtualCages: includeVirtualCages });

    // Add virtual cage units from the linear system
    for (const { cells: vcells, total: vtotal, distinct, precomputedSolns: precompSolns } of includeVirtualCages ? this.linearSystem.virtualCages : []) {
      const vunitId = this.units.length;
      const cells = vcells as Cell[];
      this._addUnit({ unitId: vunitId, kind: UnitKind.CAGE, cells, distinctDigits: distinct });
      if (precompSolns !== null) {
        this.cageSolns.push(precompSolns);
      } else {
        this.cageSolns.push(solSums(cells.length, 0, vtotal));
      }
    }
  }

  cageUnitId(r: number, c: number): number { return CAGE_UNIT_OFFSET + this.regions[r]![c]!; }

  override removeCandidate(r: number, c: number, d: number): BoardEvent[] {
    const events = super.removeCandidate(r, c, d);
    if (events.length === 0) return events; // d wasn't a candidate — nothing changed, nothing to prune

    // Prune cage solutions for all cage units containing this cell
    for (const uid of this.cellUnitIds(r, c)) {
      if (this.units[uid]!.kind === UnitKind.CAGE) {
        events.push(...this._pruneCageSolutions(uid - CAGE_UNIT_OFFSET, r, c, d));
      }
    }

    return events;
  }

  removeCageSolution(cageIdx: number, solution: readonly number[]): BoardEvent {
    const solns = this.cageSolns[cageIdx]!;
    const idx = solns.findIndex(s => s.length === solution.length && s.every((d, i) => d === solution[i]));
    if (idx >= 0) solns.splice(idx, 1);
    const cageUnitId = CAGE_UNIT_OFFSET + cageIdx;
    return { trigger: Trigger.SOLUTION_PRUNED, payload: cageUnitId, hintDigit: null };
  }

  private _pruneCageSolutions(cageIdx: number, _r: number, _c: number, d: number): BoardEvent[] {
    const cageUnit = this.units[CAGE_UNIT_OFFSET + cageIdx]!;
    // If d is still possible somewhere in the cage, nothing to prune
    if (cageUnit.cells.some(([cr, cc]) => this.cands(cr, cc).has(d))) return [];
    // Remove all solutions containing d
    const toRemove = this.cageSolns[cageIdx]!.filter(s => s.includes(d));
    return toRemove.map(s => this.removeCageSolution(cageIdx, s));
  }

  /**
   * Builds { cageOf, cageTotal, cageCells } from this.spec — the same
   * extraction mrvBacktrack used to perform inline (now moved here so the
   * backtracker can ask the board for its constraints generically).
   */
  override cageConstraints(): CageConstraints {
    const cageOf = Array.from({length: 9}, () => new Array<number>(9).fill(0));
    const cageTotal = new Map<number, number>();
    const cageCells = new Map<number, Cell[]>();

    for (let r = 0; r < 9; r++) {
      for (let c = 0; c < 9; c++) {
        const cid = this.spec.regions[r]![c]!; // 1-based
        cageOf[r]![c] = cid;
        if (!cageCells.has(cid)) cageCells.set(cid, []);
        cageCells.get(cid)!.push([r, c] as Cell);
        const t = this.spec.cageTotals[r]![c]!;
        if (t !== 0) cageTotal.set(cid, t);
      }
    }

    return { cageOf, cageTotal, cageCells };
  }

  /**
   * Add a user-acknowledged virtual cage as a new cage unit.
   *
   * @param cells - cells that form the cage (row-major Cell tuples)
   * @param total - the cage sum constraint
   * @param eliminatedSolns - solutions the user has already ruled out; pre-filtered from the generated solution list
   * @param distinct - whether digits must be distinct within the cage (default true)
   */
  addVirtualCage(
    cells: readonly Cell[],
    total: number,
    eliminatedSolns: readonly (readonly number[])[],
    { distinct = true, negativeCells, eliminatedDiffSolns }: {
      distinct?: boolean;
      negativeCells?: readonly Cell[];
      eliminatedDiffSolns?: readonly DiffSolution[];
    } = {},
  ): void {
    const vunitId = this.units.length;

    if (negativeCells && negativeCells.length > 0) {
      // Diff cage: populate cageSolns with combined sorted digit arrays so that
      // CageCandidateFilter and SolutionMapFilter work correctly.
      // Combined [pos ∪ neg] is sound: any digit absent from all combined solutions
      // cannot appear in any cage cell regardless of role.
      const negKeys = new Set(negativeCells.map(([r, c]) => `${r},${c}`));
      const posCount = cells.length - negKeys.size;
      const negCount = negKeys.size;
      const diffKey = (s: DiffSolution) => `${[...s.pos].join(',')}|${[...s.neg].join(',')}`;
      const elimSet = new Set((eliminatedDiffSolns ?? []).map(diffKey));
      const solns = solDiffs(posCount, negCount, total)
        .filter(s => !elimSet.has(diffKey(s)))
        .map(s => [...s.pos, ...s.neg].sort((a, b) => a - b));
      this.cageSolns.push(solns);
    } else {
      const elimSet = new Set(eliminatedSolns.map(s => s.slice().sort().join(',')));
      const solns = solSums(cells.length, 0, total)
        .filter(s => !elimSet.has(s.slice().sort().join(',')));
      this.cageSolns.push(solns);
    }

    this._addUnit({ unitId: vunitId, kind: UnitKind.CAGE, cells, distinctDigits: distinct });
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function nextInSet<T>(s: Set<T>): T {
  return s.values().next().value as T;
}

/** Validate that a fully-solved board satisfies row/col/box constraints (cages are skipped — validate separately). */
export function validateSolution(board: KillerBoardState): string[] {
  const violations: string[] = [];
  for (const unit of board.units) {
    if (unit.kind === UnitKind.CAGE) continue; // cage validation is separate
    const digits = unit.cells.map(([r, c]) => nextInSet(board.cands(r, c)));
    const uniq = new Set(digits);
    if (uniq.size !== 9 || ![...uniq].every(d => d >= 1 && d <= 9))
      violations.push(`Unit ${unit.unitId} (${UnitKind[unit.kind]}) has duplicate or invalid digits`);
  }
  return violations;
}

/** Intersection of all sets, returned as a sorted array. */
export function intersectAll(sets: ReadonlySet<number>[]): number[] {
  if (sets.length === 0) return [];
  return [...sets[0]!].filter(d => sets.every(s => s.has(d))).sort((a, b) => a - b);
}
